_posts/2019-10-11-hunting-down-a-dask-bug.markdown

```diff
@@ -0,0 +1,174 @@
+---
+layout: single
+title:  "Hunting down a Dask Bug - or What to Do if You're Randomly Losing Data"
+date:   2019-05-28 20:00:00 +0100
+tags: technology python data-engineering
+header:
+  overlay_image: assets/images/tech_gear_banner.jpg
+  overlay_filter: 0.2
+  show_overlay_excerpt: false
+author: Andreas Merkel
+author_profile: true
+---
+# Hunting down a Dask Bug - or What to Do if You're Randomly Losing Data
+
+When introducing a new technology into your stack, you have to be prepared to discover
+bugs. You are using someone else's work, and probably in a way that is, to some extent,
+different to the way others have used it before you. And this does not finish when you
+are done with your implementation and testing, but extends to the time when you
+finally put it in production.
+
+This is a story about how subtle bugs can manifest themselves after months of having
+a technology in production, and how tricky it can be to track them down.
+
+## The setup
+
+We are currently in the process of adopting
+[Dask/Distributed](https://distributed.dask.org) for distributed computing and are using
+it, among other things, to work with data in flat files on BLOB storage (cf.
+[Introducing Kartothek - Consistent parquet table management powered by Apache Arrow and
+Dask](https://tech.jda.com/introducing-kartothek/).)
+A typical use case we have for Dask/Distributed is reading a collection of flat files,
+which we call *dataset*, doing some calculation on the data and writing it again to
+another dataset.
+
+We had been gradually switching more and more of our data pipelines to Dask/Distributed
+in favor of a proprietary distributed computing framework that we want to phase out when
+things started to get a tad mysterious and data began to disappear in an unexpected way.
+
+## The incident
+
+It started when one of our customers raised an incident complaining that we did not
+provide any data for one of their stores via our API one day. We provide data to them
+that is calculated on a daily basis, and so far, everything had been OK for months after
+switching to Dask/Distributed for data processing.
+
+The API provides the data from a dataset (let's call it _output dataset_), which is
+filled from another dataset using Dask/Distributed (let's call this one _input dataset_)
+by shuffling around the data. The shuffling is done because the input dataset is
+organized in a way that reflects the customer's stores (basically one [Apache
```

Comment:
i think it's " the customers' "

Reply:
I think both variants work here. We're talking about a single customer here, but the data is organized this way for all our customers.